zmq-eio/src/socket.ml

```diff
@@ -0,0 +1,159 @@
+(** Eio based bindings for eio *)
+exception Closed
+
+type 'a t = {
+  socket : 'a Zmq.Socket.t;
+  fd : Unix.file_descr;
+  senders : (unit -> unit) Queue.t;
+  receivers : (unit -> unit) Queue.t;
+  condition : Eio.Condition.t;
+  mutex : Eio.Mutex.t;
+  ready_condition: Eio.Condition.t;
+  mutable thread : unit Eio.Promise.or_exn option; (* None indicates already closed *)
+}
+
+type 'a of_socket_args = sw:Eio.Switch.t -> 'a
+type 'a deferred = 'a
+
+(** invoke the first function on the queue, but only pop it if it does not raise EAGAIN *)
+let process queue =
+  match (Queue.peek queue) () with
+  | () ->
+    let (_: unit -> unit) = Queue.pop queue in
+    ()
+  | exception Unix.Unix_error (Unix.EAGAIN, _, _) ->
+    (* If f raised EAGAIN, dont pop the message. *)
+    (* This should never happen. If so, the queue could be replaced with a Eio.Stream for faster handling *)
+    ()
+
+let with_lock lock f =
+  Eio.Mutex.lock lock;
+  try
+    let v = f () in
+    Eio.Mutex.unlock lock;
+    v
+  with
+  | e ->
+    Eio.Mutex.unlock lock;
+    raise e
+
+let rec fd_monitor t =
+  Eio.Condition.await_no_mutex t.ready_condition;
+  Eio_unix.await_readable t.fd;
+  with_lock t.mutex (fun () -> Eio.Condition.broadcast t.condition);
+  fd_monitor t
+
+let rec event_loop t =
+  let inner () =
+    match Zmq.Socket.events t.socket with
+    | Zmq.Socket.Poll_error ->
+      failwith "Cannot poll socket"
+    | (Poll_in_out | Poll_in) when not (Queue.is_empty t.receivers) ->
+      process t.receivers
+    | (Poll_in_out | Poll_out) when not (Queue.is_empty t.senders) ->
+      process t.senders
+    | _ ->
+      Eio.Condition.broadcast t.ready_condition;
+      Eio.Condition.await t.condition t.mutex;
+  in
+  with_lock t.mutex (fun () -> inner ());
+  match t.thread with
+  | None when Queue.is_empty t.senders && Queue.is_empty t.receivers ->
+    ()
+  | _ ->
+    event_loop t
+
+let of_socket: ('a Zmq.Socket.t -> 'a t) of_socket_args = fun ~sw socket ->
+  let fd = Zmq.Socket.get_fd socket in
+  let t =
+    { socket;
+      fd;
+      senders = Queue.create ();
+      receivers = Queue.create ();
+      mutex = Eio.Mutex.create ();
+      condition = Eio.Condition.create ();
+      ready_condition = Eio.Condition.create ();
+      thread = None;
+    }
+  in
+  let thread = Eio.Fiber.fork_promise ~sw (fun () ->
+    Eio.Switch.run (fun sw ->
```

Comment:
I don't understand the Eio terms here, what does the `Switch` do?

Reply:
Fibers is a lightweight thread. A switch is a collection of fibers and will run until all fibers has ended or cancelled. Daemon fibers will automatically be cancelled once all "regular" fibers has ended. 

The construct here is made to control lifetime of the two fibers started. Once the main loop exits, the signalling daemon fiber will be automatically be cancelled. Creating a promise allows the caller of `close` to wait until the event loop has ended (all handlers completed).

Close is meant to be a nice way to close the socket allowing all readers and writers to end. Its also possible to cancel a switch. If the parent switch (passed as arg this function) is cancelled all child fibers and switches are cancelled. 

I hope this explains the construct of the fork -> switch -> fork.